sdk/offchain/src/payment_command.rs

```diff
@@ -0,0 +1,283 @@
+        my_actor: Actor,
+        cid: Uuid,
+        prior: Option<&Self>,
+    ) -> Result<Self, &'static str> {
+        let payment_state = Self::validate(origin, my_actor, &payment, prior)?;
+
+        Ok(Self {
+            payment,
+            payment_state,
+            origin,
+            my_actor,
+            cid,
+        })
+    }
+
+    pub fn payment(&self) -> &PaymentObject {
+        &self.payment
+    }
+
+    pub fn origin(&self) -> Origin {
+        self.origin
+    }
+
+    pub fn cid(&self) -> Uuid {
+        self.cid
+    }
+
+    pub fn reference_id(&self) -> Uuid {
+        self.payment().reference_id()
+    }
+
+    pub fn my_actor(&self) -> Actor {
+        self.my_actor
+    }
+
+    pub fn counterparty_actor(&self) -> Actor {
+        self.my_actor.counterparty_actor()
+    }
+
+    pub fn counterparty_actor_object(&self) -> &PaymentActorObject {
+        self.payment
+            .actor_object_by_actor(self.counterparty_actor())
+    }
+
+    pub fn my_actor_object(&self) -> &PaymentActorObject {
+        self.payment.actor_object_by_actor(self.my_actor())
+    }
+
+    pub fn payment_state(&self) -> PaymentState {
+        self.payment_state
+    }
+
+    pub fn follow_up_action(&self) -> Option<PaymentAction> {
+        PaymentAction::follow_up_action(self.payment_state(), self.my_actor())
+    }
+
+    pub fn to_request(&self) -> CommandRequestObject {
+        let command = Command::PaymentCommand(PaymentCommandObject::new(self.payment.clone()));
+        CommandRequestObject::new(command, self.cid())
+    }
+
+    fn validate(
+        origin: Origin,
+        my_actor: Actor,
+        payment_object: &PaymentObject,
+        prior: Option<&PaymentCommand>,
+    ) -> Result<PaymentState, &'static str> {
+        let payment_state =
+            PaymentState::from_payment(payment_object).ok_or("invalid payment state")?;
+        // Validate state trigger actor
+        if origin.is_inbound() && my_actor.counterparty_actor() != payment_state.trigger_actor() {
+            return Err("Should not produce");
+        }
+
+        if let Some(prior) = prior {
+            // Does the prior command have the same reference_id?
+            if payment_object.reference_id() != prior.payment().reference_id() {
+                return Err("payment object reference_id does not match");
+            }
+
+            // Validate actor object
+            if origin.is_inbound()
+                && payment_object.actor_object_by_actor(my_actor) != prior.my_actor_object()
+            {
+                return Err("invalid overwrite");
+            }
+
+            // Validate WriteOnce fields
+            payment_object
+                .validate_write_once_fields(prior.payment())
+                .map_err(|_| "write once fields error")?;
+
+            // Validate transition
+            if !PaymentState::is_valid_transition(prior.payment_state(), payment_state) {
+                return Err("invalid state transition");
```

Comment:
need to read the code more, but could this be due to a replay of a previous message, should the response be a replay of the success or an error?